---
title: "RNA-seq low level EDA"
output:
  html_document:
  toc: true
---

# Goals of this workshop

* Low-level EDA (exploratory data analysis) of RNA-seq
* Examine biases affecting paired-end fragments

We will not cover, but you can read about elsewhere:

* Counting RNA-seq reads/fragments in genes
  * Using, e.g. featureCounts, htseq, or summarizeOverlaps
  * Or tximport following ultra-fast quantification software, e.g. Sailfish, Salmon, kallisto
* [Basic differential expression analysis workflow](http://www.bioconductor.org/help/workflows/rnaseqGene/)

# Gene annotation

We start by looking at two genes, USF2 and CHPF. Let's obtain some information on these genes from the *Homo.sapiens* package. We will pull out the 
transcript ID and name that go along with the gene symbols.

```{r}
library(Homo.sapiens)
columns(Homo.sapiens)
g <- list()
g[["USF2"]] <- select(Homo.sapiens, "USF2", c("TXID","TXNAME"), "SYMBOL")
g[["CHPF"]] <- select(Homo.sapiens, "CHPF", c("TXID","TXNAME"), "SYMBOL")
g
```

We can use a *TxDb* object to get the exons for the transcripts
of these two genes. The *exonsBy* function returns a *GRangesList*
object, and by specifying `by="tx"`, within this *GRangesList* 
we can find the GRanges for each exon of every transcript.
In `ebt` we have the exons of every transcript, labelled by
transcript ID. We then use the information from the previous
chunk to pull out the transcripts of USF2 and CHPF:

```{r}
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene
ebt <- exonsBy(txdb, by="tx")
head(names(ebt))
usf2 <- ebt[ g[["USF2"]]$TXID ]
chpf <- ebt[ g[["CHPF"]]$TXID ]
```

We can take a look at these:

```{r}
usf2
chpf
```

It will be useful to calculate the total range of the 
exons of the transcript of each gene:

```{r}
usf2.r <- range(unlist(usf2))
chpf.r <- range(unlist(chpf))
```

Let's now visualize the exons of the transcripts using base R.
First, let's write a line of code that draws the exons
of a single transcript of USF2

```{r}
library(rafalib)
bigpar()
lens <- length(usf2)
nullplot(start(usf2.r),end(usf2.r),0,2)
segments(start(usf2[[1]]), rep(1,lens[1]), 
           end(usf2[[1]]), rep(1,lens[1]), lwd=3)
```

Now, to avoid repeating this code over and over for each
gene, let's write a general function for drawing the 
exons of each transcript of a gene, 
contained in a `GRangeList` x.

```{r}
plotGRangesList <- function(x,name) {
  r <- range(unlist(range(x)))
  nullplot(start(r),end(r),0.5,length(x)+0.5,
           main=name,xlab=seqnames(x[[1]][1]))
  lens <- elementNROWS(x)
  for (i in seq_along(x)) {
    segments(start(x[[i]]), rep(i,lens[i]), 
             end(x[[i]]), rep(i,lens[i]), lwd=3)
  }
}
plotGRangesList(usf2, "USF2")
plotGRangesList(chpf, "CHPF")
```

```{r}
library(GenomicAlignments)
bamfiles <- list.files("bams",full=TRUE)
sortBam(bamfiles[1], destination="sorted")
indexBam("sorted.bam")
usf2.r <- keepStandardChromosomes(usf2.r)
gap <- readGAlignmentPairs("sorted.bam", param=ScanBamParam(which=usf2.r))
```

```{r}
cov <- coverage(gap)
# rename this cov.genome
cov.genome <- as.integer(cov[usf2.r][["chr19"]])
pos.genome <- seq(from=start(usf2.r),to=end(usf2.r))
plot(pos.genome, cov.genome, xlab="position (genome)", ylab="coverage")
```

```{r}
m2tx <- mapToTranscripts(as(gap,"GRanges"), usf2)
cov.tx <- coverage(m2tx)
bigpar(3,2)
for (i in seq_along(usf2)) {
  plot(as.integer(cov.tx[[i]]), xlab="position (tx)", 
       ylab="coverage", main=names(usf2)[i])
}
```

